### PR TITLE
Added controller trainer and improved truck class

### PR DESCRIPTION
**Pull request by dafaronbi**
### Added

- new truck methods for randomizing state within contraints
- new truck methods for seeing if truck is at dock or offscreen
- Training script for optimizing controller 

### Note
I currently have not successfully trained the controller to convergence. I have based the training off of [this](https://tifu.github.io/truck_backer_upper/). On the website, they mention that the controller is hard to train. I have tried training it on the website with no success, so it seems like even their lessons are difficult to train. However, the code for training should be very similar to the code on the website. You may also alter the amount of lessons, max time steps, learning rate etc. to see if the model converges. I have been trying for over a week and have not succeeded yet.

### CODE REVIEW

File: 14-truck_backer-upper.ipynb [hunk 1]
```diff
@@ -243,16 +1290,24 @@
...
+     ]
+    }
+   ],
    "source": [
     "# Build expert data set\n",
     "\n",
     "episodes = 10\n",
     "inputs = list()\n",
     "outputs = list()\n",
-    "# truck = Truck(); episodes = 10_000  # uncooment for creating the data set\n",
```
Comment (Atcold) [inline 14-truck_backer-upper.ipynb:255]: Don't change comments into executable instructions.

File: 14-truck_backer-upper.ipynb [hunk 2]
```diff
@@ -268,16 +1323,27 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 9,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Comment (Atcold) [inline 14-truck_backer-upper.ipynb:1328]: You forgot to clean the output of every cell.

File: 14-truck_backer-upper.ipynb [hunk 3]
```diff
@@ -335,18 +1409,516 @@
...
     "len(train_inputs)"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 15,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Comment (Atcold) [inline 14-truck_backer-upper.ipynb:1434]: Why pushing all these numbers?

File: 14-truck_backer-upper.ipynb [hunk 4]
```diff
@@ -387,20 +1967,305 @@
...
+    "criterion = nn.MSELoss()\n",
+    "\n",
+    "#parameters for training lessons\n",
+    "#CHANGE THESE TO SUITE YOUR NEEDS\n",
+    "num_lessons = 20\n",
+    "num_samples =1_000\n",
+    "max_steps = 1_000\n",
+    "\n",
+    "#define lessons\n",
+    "x_min = np.linspace(4, 30, num_lessons)\n",
```
Comment (Atcold) [inline 14-truck_backer-upper.ipynb:2137]: We use `torch`, not `numpy`.

File: 14-truck_backer-upper.ipynb [hunk 5]
```diff
@@ -387,20 +1967,305 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
   }
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3 [conda env:pDL]",
```
Comment (Atcold) [inline 14-truck_backer-upper.ipynb:403]: The class's kernel is `pDL`, not `ipykernel`.

**Review by Atcold — CHANGES_REQUESTED** [review]
There's a lot of stuff I'm confused about…

File: 14-truck_backer-upper.ipynb [hunk 6]
```diff
@@ -92,12 +119,12 @@
     "        \n",
     "        # Check for illegal conditions\n",
     "        if self.is_jackknifed():\n",
-    "            print('The truck is jackknifed!')\n",
-    "            return\n",
+    "            # print('The truck is jackknifed!')\n",
```
[the remark this answers is not in the document]
Comment (dafaronbi) [reply 14-truck_backer-upper.ipynb:122]: This can be uncommented if you like, It was just removed so it is not constantly printed during training. Same with the other comment on the car being off screen you posted.

File: 14-truck_backer-upper.ipynb (hunk 1 above)
Comment (dafaronbi) [reply 14-truck_backer-upper.ipynb:255]: I'm confused? The comment said  I should un-comment when creating data set
